macaw-riscv-symbolic/src/Data/Macaw/RISCV/Symbolic.hs

```diff
@@ -0,0 +1,299 @@
+
+-- | The RISC-V control/status registers that are directly supported by Macaw.
+csrRegs :: Ctx.Assignment (MR.RISCVReg rv) (RR.CtxRepeat 23 (MT.BVType (G.RVWidth rv)))
+csrRegs = RR.repeatAssign (MR.CSR . toEnum)
+
+-- | This contains an assignment that stores the register associated with each index in the
+-- RISC-V register structure.
+riscvRegAssignment :: Ctx.Assignment (MR.RISCVReg rv) (MS.ArchRegContext (MR.RISCV rv))
+riscvRegAssignment =
+  Ctx.Empty Ctx.:> MR.PC
+  Ctx.<++> gprRegs
+  Ctx.<++> fprRegs
+  Ctx.<++> csrRegs
+  Ctx.<++> (Ctx.Empty Ctx.:> MR.PrivLevel Ctx.:> MR.EXC)
+
+riscvRegStructType ::
+     forall rv
+   . G.KnownRV rv
+  => G.RVRepr rv
+  -> C.TypeRepr (MSB.ArchRegStruct (MR.RISCV rv))
+riscvRegStructType _rv =
+  C.StructRepr $ MS.typeCtxToCrucible $ FC.fmapFC MT.typeRepr $ riscvRegAssignment @rv
+
+regIndexMap :: forall rv
+             . G.KnownRV rv
+            => MSB.RegIndexMap (MR.RISCV rv)
+regIndexMap = MSB.mkRegIndexMap assgn sz
+  where
+    assgn = riscvRegAssignment @rv
+    sz = Ctx.size $ MS.typeCtxToCrucible $ FC.fmapFC MT.typeRepr assgn
+
+lookupReg :: forall rv m f tp
+           . (G.KnownRV rv, Typeable rv, X.MonadThrow m)
+          => MR.RISCVReg rv tp
+          -> Ctx.Assignment f (MS.MacawCrucibleRegTypes (MR.RISCV rv))
+          -> m (f (MS.ToCrucibleType tp))
+lookupReg r asgn =
+  case MapF.lookup r regIndexMap of
+    Nothing -> X.throwM (MissingRegisterInState (Some r))
+    Just pair -> return (asgn Ctx.! MSB.crucibleIndex pair)
+
+archLookupReg :: (G.KnownRV rv, Typeable rv)
+              => C.RegEntry sym (MS.ArchRegStruct (MR.RISCV rv))
+              -> MR.RISCVReg rv tp
+              -> C.RegEntry sym (MS.ToCrucibleType tp)
+archLookupReg regEntry reg =
+  case lookupReg reg (C.regValue regEntry) of
+    Just (C.RV val) -> C.RegEntry (MS.typeToCrucible $ MT.typeRepr reg) val
+    Nothing -> error $ "unexpected register: " ++ show (MC.prettyF reg)
+
+updateReg :: forall rv m f tp
+           . (G.KnownRV rv, Typeable rv, X.MonadThrow m)
+          => MR.RISCVReg rv tp
+          -> (f (MS.ToCrucibleType tp) -> f (MS.ToCrucibleType tp))
+          -> Ctx.Assignment f (MS.MacawCrucibleRegTypes (MR.RISCV rv))
+          -> m (Ctx.Assignment f (MS.MacawCrucibleRegTypes (MR.RISCV rv)))
+updateReg r upd asgn = do
+  case MapF.lookup r regIndexMap of
+    Nothing -> X.throwM (MissingRegisterInState (Some r))
+    Just pair -> return (asgn & MapF.ixF (MSB.crucibleIndex pair) %~ upd)
+
+archUpdateReg :: (G.KnownRV rv, Typeable rv)
+              => C.RegEntry sym (MS.ArchRegStruct (MR.RISCV rv))
+              -> MR.RISCVReg rv tp
+              -> C.RegValue sym (MS.ToCrucibleType tp)
+              -> C.RegEntry sym (MS.ArchRegStruct (MR.RISCV rv))
+archUpdateReg regEntry reg val =
+  case updateReg reg (const $ C.RV val) (C.regValue regEntry) of
+    Just r -> regEntry { C.regValue = r }
+    Nothing -> error $ "unexpected register: " ++ show (MC.prettyF reg)
```

Comment:
I would expect this to be a "this should never happen" situation, in which case I would recommend using `panic` which provides more information for debugging.

Reply:
It is possible to observe this `error` if you call `archUpdateReg` with RISC-V's `x0` register, which we intentionally do not include in the `regIndexMap` (see the comments for `gprRegs` above). I suppose there is a philosophical question of whether we should:

1. Anticipate this situation and throw an `error` message here, or
2. Document that this function has a precondition that the `RISCVReg` argument must belong to the set of registers that `macaw-riscv-symbolic` tracks, and panic if this precondition is violated

Personally, I lean more towards option (1), and that's how other `macaw-*-symbolic` backends currently implement their own versions of `archUpdateReg`. I could be talked into implementing (2) if you prefer that, however.

Reply:
I suppose my concern is that if you do get an "unexpected register" error and it's not caught and handled elsewhere that it's hard to track down and even when you do, the context of what is wrong takes some time to deduce.  I'm not clear from your two choices why #1 would use `error` and #2 would use `panic`.  I'm somewhat mollified by the argument that this fits the trend elsewhere, but I also wonder if the trend was well thought out or was a case of "I guess I need something here so the compiler doesn't complain about a missing case".  I trust your instincts on the best way to handle this though.

Reply:
> I'm not clear from your two choices why https://github.com/GaloisInc/macaw/issues/1 would use `error` and https://github.com/GaloisInc/macaw/issues/2 would use `panic`.

My reasoning is that it's not just a missing case: it's genuinely possible for a `macaw-riscv-symbolic` API user to trigger this error if they are not careful when calling it. Whether that deserves an `error` or `panic` is a matter of taste, I suppose, since although it's not _impossible_ to trigger this error, it's nevertheless very unlikely.

Reply:
I've converted these to panics. @kquick has opened https://github.com/GaloisInc/macaw/issues/412 to track the remaining task of changing the return types of `GenArchVals`' `{lookup,update}Reg` functions in `macaw-symbolic`, which are ultimately the reason why these functions have to be partial in the first place.